pkg/controller/registry/resolver/resolver.go

```diff
@@ -1,222 +1,567 @@
-				subStep, err := NewSubscriptionStepResource(namespace, *op.SourceInfo())
-				if err != nil {
-					return nil, nil, nil, err
+	depIds := make([]solver.Identifier, 0)
+	for _, c := range candidates {
+		// track which operator this is replacing, so that it can be realized when creating the resources on cluster
+		if current != nil {
+			c.Replaces = current.Identifier()
+		}
+		depIds = append(depIds, c.Identifier())
+	}
+
+	// all candidates added as options for this constraint
+	subInstallable.AddDependency(depIds)
+
+	return installables, nil
+}
+
+func (r *SatResolver) getBundleInstallables(catalog registry.CatalogKey, predicates []OperatorPredicate, namespacedCache MultiCatalogOperatorFinder, visited map[OperatorSurface]*BundleInstallable) (map[solver.Identifier]struct{}, map[solver.Identifier]*BundleInstallable, error) {
+	errs := make([]error, 0)
+	installables := make(map[solver.Identifier]*BundleInstallable, 0) // all installables, including dependencies
+
+	var finder OperatorFinder = namespacedCache
+	if !catalog.Empty() {
+		finder = namespacedCache.Catalog(catalog)
+	}
+
+	bundleStack := finder.Find(predicates...)
+
+	// track the first layer of installable ids
+	var initial = make(map[*Operator]struct{})
+	for _, o := range bundleStack {
+		initial[o] = struct{}{}
+	}
+
+	for {
+		if len(bundleStack) == 0 {
+			break
+		}
+		// pop from the stack
+		bundle := bundleStack[len(bundleStack)-1]
+		bundleStack = bundleStack[:len(bundleStack)-1]
+
+		bundleSource := bundle.SourceInfo()
+		if bundleSource == nil {
+			err := fmt.Errorf("unable to resolve the source of bundle %s, invalid cache", bundle.Identifier())
+			errs = append(errs, err)
+			continue
+		}
+
+		if b, ok := visited[bundle]; ok {
+			installables[b.identifier] = b
+			continue
+		}
+
+		bundleInstallable := NewBundleInstallable(bundle.Identifier(), bundle.Channel(), bundleSource.Catalog)
+		visited[bundle] = &bundleInstallable
+
+		dependencyPredicates, err := bundle.DependencyPredicates()
+		if err != nil {
+			errs = append(errs, err)
+			continue
+		}
+		for _, d := range dependencyPredicates {
+			// errors ignored; this will build an empty/unsatisfiable dependency if no candidates are found
+			candidateBundles, _ := AtLeast(1, namespacedCache.FindPreferred(&bundle.sourceInfo.Catalog, d))
+			sortedBundles, err := r.sortBundles(candidateBundles)
+			if err != nil {
+				errs = append(errs, err)
+				continue
+			}
+			bundleDependencies := make([]solver.Identifier, 0)
+			for _, dep := range sortedBundles {
+				found := namespacedCache.Catalog(dep.SourceInfo().Catalog).Find(WithCSVName(dep.Identifier()))
+				if len(found) > 1 {
+					err := fmt.Errorf("found duplicate entries for %s in %s", bundle.Identifier(), dep.sourceInfo.Catalog)
+					errs = append(errs, err)
+					continue
 				}
-				steps = append(steps, &v1alpha1.Step{
-					Resolving: name,
-					Resource:  subStep,
-					Status:    v1alpha1.StepStatusUnknown,
-				})
+				b := found[0]
```

Comment:
This is not guarded for `found` length to be less than 1.

Reply:
`candidateBundles, _ := AtLeast(1, namespacedCache.FindPreferred(&bundle.sourceInfo.Catalog, d))`

The `AtLeast(1, <>)` ensures that more than result is returned, so it will error out before it gets here

Reply:
Oh wrong query - Good catch!